rfcs/20250910-graph-api-dropout/README.md

```diff
@@ -0,0 +1,183 @@
+generate the dropout mask differs by device:
+
+- CPU: Uses torch.Generator backed by the C++ Mersenne Twister engine [[#3]][3],
+  which consumes a 64-bit seed.
+- CUDA / XPU: Uses a 64-bit Philox RNG state (seed + offset) obtained from the
+  device generator [[#4]][4][[#5]][5]. The offset is advanced after each kernel
+  launch to carve out disjoint subsequences of random numbers.
+
+**TensorFlow**
+
+tf.keras.layers.Dropout [[#6]][6] takes `rate` (probability), an optional `noise_shape`
+(shape for randomly generated keep/drop flags) and an optional seed as
+attributes, with one input and one output. Internally, the random number
+generator produces a 64-bit pair (key, counter) from the seed and passes
+them to a selected RNG algorithm to generate the mask.
+Supported algorithms [[#7]][7] include Philox, ThreeFry, or AUTO_SELECT
+(which lets TensorFlow choose the best algorithm for the current device;
+AUTO_SELECT may pick an algorithm other than Philox or ThreeFry).
+
+**ONNX**
+
+ONNX's Dropout operation [[#8]][8] takes `data`, `ratio` (probability) and `training_mode`
+as inputs, and `seed` as an attribute. Outputs include the dropped tensor and
+the dropout mask. ONNX does not specify the algorithm used to generate random numbers.
+
+**cuDNN**
+
+cuDNN Graph library does not provide a dedicated Dropout operation. Instead, it
+introduces an RNG operation, which can be used to implement Dropout with two
+pointwise Multiply operations.
+The RNG operation takes `seed`, `offset` as inputs, supports various
+distributions (Bernoulli, Uniform, Normal) by setting attributes, and outputs
+the generated tensor. The RNG is based on the Philox algorithm.
+
+**oneDNN**
+
+In oneDNN primitive, Dropout is implemented as a binary post-op applied to
+output tensor before other post-ops. It uses the Philox algorithm and takes
+`seed` (s32) and `probability` (f32) as inputs, and outputs a `mask` (u8).
+
+## Proposal
+
+Based on the investigation above, we propose to define the Dropout operation in
+the oneDNN Graph API as follows:
+
+**Operation Kind:**
+
+- `Dropout` (C++), `dnnl_graph_op_dropout` (C)
+
+**Inputs:**
+
+- `src`: input tensor (`f32` / `bf16` / `f16`)
+- `seed`: random seed (`uint64_t`, single value)
+- `offset`: random offset (`uint64_t`, single value)
+- `p`: dropout probability (`f32`, single value)
+
+**Outputs:**
+
+- `dst`: output tensor (`f32` / `bf16` / `f16`)
+- `mask`: mask tensor (`u8`, optional)
+
+**Notes:**
+
+- Dropout will use the Philox algorithm for random number generation.
+- Data types for `src`, `p`, `dst` and `mask` are consistent with the
+  current oneDNN primitive design.
+- The `mask` output is optional, allowing flexibility for scenarios where
+  the mask is not needed.
+- The `offset` input is introduced to ensure reproducibility and correct
+  partitioning across devices and threads, following best practices from PyTorch
+  and cuDNN.
+- Data types for `seed` and `offset` are aligned to PyTorch to ensure
```

Comment:
How about the data types of probability and mask in PyTorch?

Reply:
probability is float32,
for mask, the aten sdpa doesn't define an output of mask. the aten dropout has this output, it supports u8 and boolean dtype of mask.

Reply:
Since you explicitly mentioned that `Data types for seed and offset are aligned to PyTorch`. It makes me wondering if probability and mask are not aligned.

> the aten dropout has this output, it supports u8 and boolean dtype of mask.

Better to mention this in the RFC as well and why we choose `u8`, not `boolean`. What if we are asked to support `boolean` mask?